### PR TITLE
[WIP] Runtime : Add support for Z

### LINKED ISSUE #666: Add support for s390x
This issue will track Z support for Kata.

### PR DESCRIPTION
**Pull request by ydjainopensource**
Initial support for Z. Does not pass some unit tests

Fixes #666

Signed-off-by: Yash Jain <ydjainopensource@gmail.com>

### CODE REVIEW

File: arch/s390x-options.mk [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2018 Yash Jain
+# #
+# # SPDX-License-Identifier: Apache-2.0
+# #
+# # IBM Z mainframe s390x settings
+
+MACHINETYPE := s390-ccw-virtio 
+KERNELPARAMS :=
+MACHINEACCELERATORS :=
+KERNELTYPE := uncompressed # Not sure ablout this
```
Comment (jodh-intel) [inline arch/s390x-options.mk]: I'd remove this entry then for now to get the default kernel type.

File: cli/kata-check_s390x.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	}
+	count, err := checkKernelModules(details.requiredKernelModules, archKernelParamHandler)
+	if err != nil {
+		return err
+
+	}
+	if count == 0 {
+		return nil
+
```
Comment (jodh-intel) [inline cli/kata-check_s390x.go]: Nit: extraneous blank line.

File: cli/kata-check_s390x.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	return fmt.Errorf("ERROR: %s", failMessage)
```
Comment (jodh-intel) [inline cli/kata-check_s390x.go]: Nit: missing blank line before `return`.

File: cli/kata-check_s390x_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	devNull, err := os.OpenFile(os.DevNull, os.O_WRONLY, 0666)
+	assert.NoError(err)
+	defer devNull.Close()
+
+	savedLogOutput := kataLog.Logger.Out
+
+	// discard normal output
+	//kataLog.Logger.Out = devNull
```
Comment (jodh-intel) [inline cli/kata-check_s390x_test.go]: Commented out line please remove this or uncomment.

File: cli/kata-check_s390x_test.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+				"parameter": 123,
+			},
+			"foo",
+			false,
+		},
+
+		{
+			false,
+			logrus.Fields{
+				"parameter": "unrestricted_guest",
```
Comment (jodh-intel) [inline cli/kata-check_s390x_test.go]: This is an `x86_64`-specific feature afaik.

File: cli/kata-check_test.go [hunk 6]
```diff
@@ -121,10 +122,17 @@ func makeCPUInfoFile(path, vendorID, flags string) error {
 	if err != nil {
 		return err
 	}
-
-	args := map[string]string{
-		"Flags":    flags,
-		"VendorID": vendorID,
+	var args map[string]string
+	if goruntime.GOARCH == "s390x" {
```
Comment (jodh-intel) [inline cli/kata-check_test.go]: I'd rather we don't add architecture-specific checks like this and instead do one of:

- Make `makeCPUInfoFile()` architecture-specific.
  - Rename this function `genericMakeCPUInfoFile()`.
  - Add a `makeCPUInfoFile()` to each `kata-check_${GOARCH}_test.go` file.
  - Make the `makeCPUInfoFile()` functions call `genericMakeCPUInfoFile()`, but for 390 have a custom implementation.
- Define the `args` map as in each `kata-check_${GOARCH}_test.go` file. Again, we could make each architecture use a generic map or they can provide their own custom implementation.

File: cli/oci.go [hunk 7]
```diff
@@ -251,10 +252,16 @@ func isCgroupMounted(cgroupPath string) bool {
 	if err := syscall.Statfs(cgroupPath, &statFs); err != nil {
 		return false
 	}
+	if goruntime.GOARCH == "s390x" {
```
Comment (jodh-intel) [inline cli/oci.go]: Again, this code should live in `cli/oci_s390.go`. Same comment for other `GOARCH` checks below.

File: cli/oci.go [hunk 8]
```diff
@@ -251,10 +252,16 @@ func isCgroupMounted(cgroupPath string) bool {
...
+		if statFs.Type != uint32(cgroupFsType) {
+			return false
+		}
 
-	if statFs.Type != int64(cgroupFsType) {
-		return false
-	}
+	} // else {
```
Comment (jodh-intel) [inline cli/oci.go]: Please remove/uncomment this commented out block.

File: virtcontainers/hypervisor.go [hunk 9]
```diff
@@ -505,6 +505,10 @@ func RunningOnVMM(cpuInfoPath string) (bool, error) {
 
 	flagsField := "flags"
 
+	if runtime.GOARCH == "s390x" {
+		flagsField = "features"
+
```
Comment (jodh-intel) [inline virtcontainers/hypervisor.go]: Nit: extraneous blank line.

File: virtcontainers/qemu.go [hunk 10]
```diff
@@ -89,6 +89,7 @@ var qemuMinorVersion int
 // agnostic list of kernel parameters
 var defaultKernelParameters = []Param{
 	{"panic", "1"},
+	{"initcall_debug", ""},
```
Comment (jodh-intel) [inline virtcontainers/qemu.go]: OOI, why are you re-adding this?

File: virtcontainers/qemu.go [hunk 11]
```diff
@@ -1205,6 +1206,10 @@ func genericBridges(number uint32, machineType string) []Bridge {
 		bt = pciBridge
 	case QemuPseries:
 		bt = pciBridge
+	case QemuSseries:
+		// see https://wiki.qemu.org/Documentation/Platforms/S390X#A_note_on_PCI_support
```
Comment (jodh-intel) [inline virtcontainers/qemu.go]: Nice comment! :smile:

File: virtcontainers/qemu_s390x.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+const defaultQemuPath = "/usr/bin/qemu-system-s390x"
+
+const defaultQemuMachineType = QemuSseries
+
+const defaultQemuMachineOptions = "accel=kvm,usb=off"
+
+// Not used
+const defaultPCBridgeBus = ""
+
+const defaultMemMaxS390x = 33280 // need to check
```
Comment (jodh-intel) [inline virtcontainers/qemu_s390x.go]: Unused variable so this can be removed maybe?

**Comment by ydjainopensource** [comment]
Closing this as govmm#43 is not yet vendored into runtime.

**Comment by jodh-intel** [comment]
Hi @ydjainopensource - you don't need to close - simply add the `do-not-merge` label and wait for https://github.com/intel/govmm/pull/43 to be merged. That way, we can keep reviewing this PR whilst the dependent PR is being looked at. Do you want to reopen this one?

**Comment by jodh-intel** [comment]
I'll re-open and put the label on as we don't want to lose sight of this important PR imho... ;)

**Comment by ydjainopensource** [comment]
@jodh-intel I am closing this PR now.
Unfortunately, I am at the end of my internship. Hence, I have been asked not to work on this (Z specific stuff) anymore. Someone from IBM would be taking over my work. I would personally like to have this feature added at the earliest. Mainframes and Kata would be a interesting match.

Thanks a lot to everyone in the community who made my work fun. Looking forward to working with you guys again.

**Comment by jodh-intel** [comment]
Hi @ydjainopensource - thanks *very* much for your contributions!!

I totally agree - cannot wait to see Kata running on Z, so I await IBM picking up where you left off eagerly.

Best of luck and hope to see you around :smile:

**Pull request closed without merging**